The multi platform doesn't implement `Clone`

Somewhat annoyingly, the multi platform derives `Clone`, but because of the definition:
```rust
#[derive(Clone)]
pub enum Connection<Def, Alt> where Def: DeviceInterface,
                                    Alt: DeviceInterface,
...
```
this implements `Clone` when `Def: Clone` and `Alt: Clone` but what's wanted is:
```rust
impl<Def, Alt> Clone for Connection<Def, Alt> where
    Def: DeviceInterface,
    Alt: DeviceInterface,
    Def::Connection: Clone,
    Alt::Connection: Clone,
```

```
error[E0599]: no method named `clone` found for type `surfman::platform::generic::multi::connection::Connection<surfman::platform::unix::wayland::device::Device, surfman::platform::unix::x11::device::Device>` in the current scope
  --> /home/ajeffrey/github/asajeffrey/webxr/webxr/glwindow/mod.rs:96:46
   |
96 |             let connection = self.connection.clone();
   |                                              ^^^^^ method not found in `surfman::platform::generic::multi::connection::Connection<surfman::platform::unix::wayland::device::Device, surfman::platform::unix::x11::device::Device>`
   |
   = note: the method `clone` exists but the following trait bounds were not satisfied:
           `surfman::platform::generic::multi::connection::Connection<surfman::platform::unix::wayland::device::Device, surfman::platform::unix::x11::device::Device> : std::clone::Clone`
```

--- a/surfman/src/platform/generic/multi/connection.rs
+++ b/surfman/src/platform/generic/multi/connection.rs
@@ -12,7 +12,6 @@ use super::surface::NativeWidget;
 use winit::Window;
 
 /// A connection to the display server.
-#[derive(Clone)]
 pub enum Connection<Def, Alt> where Def: DeviceInterface,
                                     Alt: DeviceInterface,
                                     Def::Connection: ConnectionInterface,
@@ -21,6 +20,20 @@ pub enum Connection<Def, Alt> where Def: DeviceInterface,
     Default(Def::Connection),
     /// The alternate connection to the display server.
     Alternate(Alt::Connection),
+}
+
+impl<Def, Alt> Clone for Connection<Def, Alt> where
+    Def: DeviceInterface,
+    Alt: DeviceInterface,
+    Def::Connection: Clone,
+    Alt::Connection: Clone,
+{
+    fn clone(&self) -> Self {
+        match self {
+	    Connection::Default(ref connection) => Connection::Default(connection.clone()),
+	    Connection::Alternate(ref connection) => Connection::Alternate(connection.clone()),
+	}
+    }
 }
 
 /// The native connection type.

--- a/surfman/src/platform/generic/multi/device.rs
+++ b/surfman/src/platform/generic/multi/device.rs
@@ -23,6 +23,20 @@ pub enum Adapter<Def, Alt> where Def: DeviceInterface, Alt: DeviceInterface {
     Alternate(<Alt::Connection as ConnectionInterface>::Adapter),
 }
 
+impl<Def, Alt> Clone for Adapter<Def, Alt> where
+    Def: DeviceInterface,
+    Alt: DeviceInterface,
+    <Def::Connection as ConnectionInterface>::Adapter: Clone,
+    <Alt::Connection as ConnectionInterface>::Adapter: Clone,
+{
+    fn clone(&self) -> Self {
+        match self {
+	    Adapter::Default(ref adapter) => Adapter::Default(adapter.clone()),
+	    Adapter::Alternate(ref adapter) => Adapter::Alternate(adapter.clone()),
+	}
+    }
+}
+
 /// A thread-local handle to a device.
 ///
 /// Devices contain most of the relevant surface management methods.

